server/src/main/java/org/elasticsearch/cluster/ClusterChangedEvent.java

```diff
@@ -336,4 +348,26 @@ private List<Index> indicesDeletedFromTombstones() {
             .toList();
     }
 
+    private static ProjectsDelta calculateProjectDelta(Metadata previousMetadata, Metadata currentMetadata) {
+        if (previousMetadata.projects().size() == 1
+            && previousMetadata.hasProject(ProjectId.DEFAULT)
+            && currentMetadata.projects().size() == 1
+            && currentMetadata.hasProject(ProjectId.DEFAULT)) {
+            return EMPTY_PROJECT_DELTA;
+        }
+
+        return new ProjectsDelta(
+            Collections.unmodifiableSet(Sets.difference(currentMetadata.projects().keySet(), previousMetadata.projects().keySet())),
+            Collections.unmodifiableSet(Sets.difference(previousMetadata.projects().keySet(), currentMetadata.projects().keySet()))
+        );
+
+    }
+
+    private static final ProjectsDelta EMPTY_PROJECT_DELTA = new ProjectsDelta(Set.of(), Set.of());
```

Comment:
nit: why not move this into the record?

Reply:
Addressed in fbe4b8188ae5a33de20e868d7c249454477f38d7